atomic_reactor/plugins/build_orchestrate_build.py

```diff
@@ -651,13 +675,161 @@ def set_build_image(self):
+            if 'from' in annotations:
+                scratch_from = json.loads(annotations['from'])
+
+                try:
+                    build_image, imagestream = self.process_image_from(scratch_from)
+                except UnknownKindException:
+                    raise RuntimeError("Build annotation has unknown 'kind' %s" %
+                                       scratch_from['kind'])
+            else:
+                raise RuntimeError("Build wasn't created from BuildConfig and neither"
+                                   " has 'from' annotation, which is needed for specified arch")
+
+        osbs = get_openshift_session(self.workflow, self.openshift_fallback)
+
+        # inspect BuildConfig
+        if build_config_name:
+            try:
+                build_config = osbs.os.get_build_config(build_config_name)
+            except OsbsException:
+                raise RuntimeError("Build config not found : %s" % build_config_name)
+
+            try:
+                build_from = build_config['spec']['strategy']['customStrategy']['from']
+            except KeyError:
+                raise RuntimeError("BuildConfig object is malformed")
+
+            try:
+                build_image, imagestream = self.process_image_from(build_from)
+            except UnknownKindException:
+                raise RuntimeError("BuildConfig object has unknown 'kind' %s" % build_from['kind'])
+
+        # if imageStream is used
+        if imagestream:
+            try:
+                tag = osbs.get_image_stream_tag(imagestream).json()
+            except OsbsException:
+                raise RuntimeError("ImageStreamTag not found %s" % imagestream)
+
+            try:
+                tag_image = tag['image']['dockerImageReference']
+            except KeyError:
+                raise RuntimeError("ImageStreamTag is malformed %s" % imagestream)
+
+            if '@sha256:' in tag_image:
+                try:
+                    labels = tag['image']['dockerImageMetadata']['Config']['Labels']
+                except KeyError:
+                    raise RuntimeError("Image in imageStreamTag '%s' is missing Labels" %
+                                       imagestream)
+
+                release = labels['release']
+                version = labels['version']
+                docker_tag = "%s-%s" % (version, release)
+                build_image = tag_image[:tag_image.find('@sha256')] + ':' + docker_tag
+            else:
+                build_image = tag_image
+
+        # we have build_image with sha, but we need tag to check for manifest list,
+        # as we want to build on different platforms
+        if build_image and '@sha256:' in build_image:
+            raise RuntimeError("Buildroot image doesn't have tag,"
+                               " which is needed for specified arch")
+
+        # we have build_image with tag, so we can check for manifest list
```

Comment:
is it worthwhile splitting this entire block off into its own function? it's clearly very self-contained.

Reply:
yeah